File: diep_frontend/src/scripts/components/obstacle.ts
import { Point, ObstacleColors, ObstacleTypes, ObstacleTypeString } from "../constants.js";
import Game from "../game.js";
import { drawObstacleLifeBar } from "../helper_services/playerDrawingHelperFunctions.js";


export default class Obstacle{
    public position: Point;
    public lifeLeft: number;
    public edges: number;
    public radius: number;
    public type: ObstacleTypeString;
    public maxLife: number;
    public id;
    public color = {
        bg: "",
        border: ""
    };

    constructor(type: ObstacleTypeString, position: Point,id:number){
        switch (type) {
            case ObstacleTypes.basic:
              this.edges = 3;
              this.color.bg = ObstacleColors.obstacle_basic_color;
              this.color.border = ObstacleColors.obstacle_basic_border_color;
              this.maxLife=5;
              break;
            case ObstacleTypes.medium:
              this.edges = 4;
              this.color.bg = ObstacleColors.obstacle_medium_color;
              this.color.border = ObstacleColors.obstacle_medium_border_color;
              this.maxLife=10;
              break;
            case ObstacleTypes.hard:
              this.color.bg = ObstacleColors.obstacle_advanced_color;
              this.color.border = ObstacleColors.obstacle_advanced_border_color
              this.edges = 5;
              this.maxLife=15;
              break;
            default:
              console.log(`No such obstacle type: ${type}`);
              break;
        }
        this.type=type;
        this.position = position;
        this.radius = 25;
        this.lifeLeft=this.maxLife;
        this.id=id;
    }

    public draw(ctx: CanvasRenderingContext2D): void {
        this.drawPolygon(ctx, this.color.border, this.radius + 10);
        this.drawPolygon(ctx, this.color.bg, this.radius);
        drawObstacleLifeBar(ctx,this);
        
    }

    public drawPolygon(ctx: CanvasRenderingContext2D, color: string, radius: number): void {
		const angle = 2*Math.PI/this.edges;
 		ctx.beginPath();
		ctx.moveTo (this.position.x +  radius*Math.cos(0), this.position.y +  radius*Math.sin(0));          
 		for (let i = 1; i <= this.edges; i++) {
			 ctx.lineTo (this.position.x + radius*Math.cos(i * angle), this.position.y + radius*Math.sin(i * angle));
		}
        ctx.fillStyle = color;
 		ctx.fill();
    }
}

File: diep_backend/game_components/obstacle.py
import math

bullet_damage = 20

class Obstacle:
    def __init__(self, position,num_edges,radius,id):
        """Class representing an obstacle in the game.

        :param   position (dict): The position of the obstacle on the game area.
        :param   num_edges (int): The number of edges the obstacle has.
        :param   radius (int): The radius of the obstacle.
        :param   id (int): The unique identifier of the obstacle.
        """
        self.position = position
        self.num_edges=num_edges
        self.num_edges += 2
        self.radius=radius+4
        self.vertices = []
        self.edges = []
        self.normals = []
        self.id=id
        if num_edges == 1:
            self.type ='basic'
            self.life_left = bullet_damage
        elif num_edges == 2:
            self.type ='medium'
            self.life_left = 2*bullet_damage
        elif num_edges == 3:
            self.type ='hard'
            self.life_left = 3*bullet_damage
        angle = 360 / self.num_edges
        for vertex_index in range(self.num_edges):
            radian_angle = math.radians(angle * vertex_index)
            x = self.radius * math.cos(radian_angle)+self.position['x']
            y = self.radius * math.sin(radian_angle)+self.position['y']
            self.vertices.append({'x': x, 'y': y})
        prev_vertex = None
        first_vertex = None
        for vertex in self.vertices:
            if prev_vertex is not None:
                x = prev_vertex['x'] - vertex['x']
                y = prev_vertex['y'] - vertex['y']
                self.edges.append({'x': x, 'y': y})
                self.normals.append({'x': -y, 'y': x})
            else:
                first_vertex = vertex
            prev_vertex = vertex
        x = first_vertex['x'] - prev_vertex['x']
        y = first_vertex['y'] - prev_vertex['y']
        self.edges.append({'x': x, 'y': y})
        self.normals.append({'x': -y, 'y': x})
